File: Task1/python_codes/q2.py
# Loay Mohamed
# Challeng 2:

exp = input("Input string to be evaluated:")
#exp = "(1+2)+(2+7)"
stack = []

for char in exp:
    if char == '(' or char == '+' or char == '-':
        stack.append(char)
    elif char == ')':
        # get the index of '('
        index = len(stack)-1
        while stack[index] != '(':
            index -= 1
        enter = False
        while stack[-1] != '(':
            #case (-num)
            if stack[index] == '(' and stack[index+1] == '-':
                num = int(stack.pop(index+2))
                stack.pop()  # pop '-'
                stack.pop()  # pop '(')
                stack.append(str(-1*num))
                break
            if not enter:
                index += 1
                enter = True
            x = int(stack.pop(index))
            if stack[-1] == '(':
                stack.pop()
                stack.append(str(x))
                break
            else:
                oper = stack.pop(index)
                y = int(stack[index])
                if oper == '+':
                    stack[index] = (str(x+y))
                else:  # oper == '-'
                    stack[index] = (str(x-y))

    elif char == ' ':
        pass
    else:
        stack.append(char)

if len(stack) > 1:
    index = 0
    while len(stack) > 1:
        x = int(stack.pop(index))
        if len(stack) == 1 and stack[-1] == '(':
            stack.pop()
            stack.append(str(x))
            break
        else:
            oper = stack.pop(index)
            y = int(stack[index])
            if oper == '+':
                stack[index] = (str(x+y))
            else:  # oper == '-'
                stack[index] = (str(x-y))
print(stack.pop())

File: Task1/kotlin_codes/q2.kt
// Loay Mohamed
// Challeng 2:

fun main(args: Array<String>) {
    
    print("Input string to be evaluated:")
    val exp = readLine() ?: ""
    //exp = "(1+2)+(2+7)"
    var stack = mutableListOf<String>();
    var char :String = ""
    for (i in 0 until exp.length ){
        char = exp[i].toString();
        if (char == "(" || char == "+" || char == "-"){
            stack.add(char)
        }
        else if (char == ")"){
            // get the index of "("
            var index = stack.size-1
            while (stack[index] != "(" ){
                index--
            }

            var enter = false
            while (stack[stack.size-1] != "(" )
            {
                //case (-num)
                if (stack[index] == "(" && stack[index+1] == "-" ){
                    val num = stack[index+2].toInt()
                    stack.removeAt(index+2)
                    stack.removeLast() // pop "-"
                    stack.removeLast() // pop "("
                    stack.add((-1*num).toString())
                    break;
                }
                if(!enter){
                    index++
                    enter = true
                }
                val x = stack[index].toInt()
                stack.removeAt(index)
                if(stack[stack.size-1] == "("){
                    stack.removeLast()
                    stack.add(x.toString())
                    break;
                }
                else {
                    val oper = stack[index]
                    stack.removeAt(index)
                    val y = stack[index].toInt()

                    if(oper == "+"){ // oper is plus
                        stack[index] = (x+y).toString()
                    }
                    else { // oper is minus
                        stack[index] = (x-y).toString()
                    }
                }
            }
        }
        else if (char == " "){
            continue;
        }
        else{
            stack.add(char)
        }
    }

    if(stack.size > 1){
        var index =0
        while(stack.size > 1){
            val x = stack[index].toInt()
            stack.removeAt(index)

            if (stack.size == 1 && stack[stack.size-1]=="("){
                stack.removeLast()
                stack.add(x.toString())
                break;
            }
            else{
                val oper = stack[index]
                stack.removeAt(index)
                val y = stack[index].toInt()

                if(oper == "+"){ // oper is plus
                    stack[index] = (x+y).toString()
                }
                else { // oper is minus
                    stack[index] = (x-y).toString()
                }
            }
        }
    }

    println(stack[0])
    stack.removeAt(0)
}